stdlib/public/core/UnsafeBufferPointer.swift.gyb

```diff
@@ -648,27 +648,329 @@ extension Unsafe${Mutable}BufferPointer {
+  /// `source.underestimatedCount`.
   ///
-  /// - Parameters:
-  ///   - repeatedValue: The instance to assign this buffer's memory to.
+  /// The returned index is the position of the next uninitialized element
+  /// in the buffer, which is one past the last element written.
+  /// If `source` contains no elements, the returned index is equal to
+  /// the buffer's `startIndex`. If `source` contains an equal or greater number
+  /// of elements than the buffer can hold, the returned index is equal to
+  /// the buffer's `endIndex`.
   ///
-  /// Warning: All buffer elements must be initialized before calling this. 
-  /// Assigning to part of the buffer must be done using the `assign(repeating:count:)`
-  /// method on the buffer’s `baseAddress`. 
+  /// - Parameter source: A sequence of elements with which to initialize the
+  ///   buffer.
+  /// - Returns: An iterator to any elements of `source` that didn't fit in the
+  ///   buffer, and an index to the next uninitialized element in the buffer.
   @inlinable // unsafe-performance
-  public func assign(repeating repeatedValue: Element) {
+  public func initialize<S: Sequence>(from source: S) -> (S.Iterator, Index)
+    where S.Element == Element {
+    return source._copyContents(initializing: self)
+  }
+
+  /// Initializes the buffer's memory with the given elements.
+  ///
+  /// Prior to calling the `initialize(fromElements:)` method on a buffer,
+  /// the memory it references must be uninitialized,
+  /// or its `Element` type must be a trivial type. After the call,
+  /// the memory referenced by the buffer up to, but not including,
+  /// the returned index is initialized.
+  ///
+  /// The returned index is the position of the next uninitialized element
+  /// in the buffer, which is one past the last element written.
+  /// If `fromElements` contains no elements, the returned index is equal to
+  /// the buffer's `startIndex`. If `fromElements` contains an equal or greater
+  /// number of elements than the buffer can hold, the returned index is equal
+  /// to the buffer's `endIndex`.
+  ///
+  /// - Parameter fromElements: A collection of elements to be used to
+  ///   initialize the buffer's storage.
+  /// - Returns: An index to the next uninitialized element in the buffer,
+  ///   or `endIndex`.
+  @discardableResult
+  @inlinable
+  @_alwaysEmitIntoClient
+  public func initialize<C: Collection>(
+    fromElements source: C
+  ) -> Index where C.Element == Element {
+    let count = Swift.min(self.count, source.count)
+    guard count > 0, let base = baseAddress else { return startIndex }
+
+    if let _ = source.withContiguousStorageIfAvailable({
+      base.initialize(
+        from: $0.baseAddress._unsafelyUnwrappedUnchecked,
+        count: count
+      )
+    }) {
+      return startIndex.advanced(by: count)
+    }
+
+    for (p, e) in zip(base..<base.advanced(by: count), source) {
+      p.initialize(to: e)
+    }
```

Comment:
I think it would probably be better to call `_copyContents` in this case -- we should avoid copying items one by one if we can help it.

(This would especially be helpful for piecewise contiguous collection types, where iterating over individual elements is pretty slow compared to iterating over the contiguous storage pieces.)

Reply:
For `_copyContents`, `self.count` must equal `source.count`. That's not necessarily satisfied in (_this version of_) this function. We could add another conditional and call it when conditions match. (Or this function gets changed.)

Reply:
D'oh, good point!

Reply:
See update.